backend/src/LibExecutionStdLib/LibList.fs

```diff
@@ -1388,6 +1388,138 @@ let fns : List<BuiltInFn> =
+
+         Preserves the order of values."
+      fn =
+        (function
+        | state, _, [ DList l; DFnVal fn ] ->
+          uply {
+            let group l =
+              let applyFn dvalA dvalB =
+                Interpreter.applyFnVal state fn [ dvalA; dvalB ]
+
+              let rec loop currentGroup groups l =
+                uply {
+                  match currentGroup, groups, l with
+                  | [], [], [] -> return Ok([])
+                  | _, _, hd :: tl ->
+                    match currentGroup with
+                    | [] -> return! loop [ hd ] groups tl
+                    | cgHd :: _ ->
+                      let! fnResult = applyFn cgHd hd
+
+                      match fnResult with
+                      | DBool true -> return! loop (hd :: currentGroup) groups tl
+                      | DBool false ->
+                        return! loop [ hd ] (List.rev currentGroup :: groups) tl
+
+                      | (DIncomplete _
+                      | DError _) as dv ->
+                        // fake dvals
+                        return Error dv
+
+                      | v ->
+                        return
+                          Exception.raiseCode (
+                            Errors.expectedLambdaType "fn" TBool v
+                          )
+                  | cg, [], [] -> return Ok(List.rev (List.rev cg :: groups))
+                  | cg, gs, [] -> return Ok(List.rev (List.rev cg :: gs))
+                }
+
+              loop [] [] l
+
+            match! group l with
+            | Ok result -> return DList(List.map (fun g -> DList g) result)
+            | Error fakeDval -> return fakeDval
+          }
+        | _ -> incorrectArgs ())
+      sqlSpec = NotYetImplemented
+      previewable = Pure
+      deprecated = NotDeprecated }
+
+
+
+    { name = fn "List" "groupBy" 0
+      typeParams = []
+      parameters =
+        [ Param.make "list" (TList varA) ""
+          Param.makeWithArgs "fn" (TFn([ varA ], varB)) "" [ "item" ] ]
+      returnType = TList(TTuple(varB, TList varA, []))
+      description =
+        "Groups <param list> into tuples (key, elements), where the key is computed by applying
+         <param fn> to each element in the list.
+
+         For example, if <param list> is {{[1, 2, 3, 4, 5]}} and <param fn>
+         is {{fn item -> Int.mod_v0 item 2}}, returns {{[(1, [1, 3, 5]), (0, [2, 4])]}}.
+
+         Preserves the order of values and of the keys."
+      fn =
+        (function
+        | state, _, [ DList l; DFnVal fn ] ->
+          uply {
+            let groupBy l =
+              let applyFn dval = Interpreter.applyFnVal state fn [ dval ]
+
+              let rec loop groups l =
+                uply {
+                  match l with
+                  | hd :: tl ->
+                    let! key = applyFn hd
+
+                    if not (Dval.isFake key) then
+                      let foundGroup = List.tryFind (fun (k, _) -> k = key) groups
```

Comment:
I think this could be done in one pass. Right now it finds the key, then goes through to do an update. Couldn't we go through and do the check and update at the same time.

Alternatively, it feels like a Map would be a better structure for this, though you would lose the order of tuples. However, I'm not super sure the ordering of tuples is important here?

Reply:
> I'm not super sure the ordering of tuples is important here
Even if it adds a little more implementation complexity, I think it's worth the quality of life to have a defined behavior for this.
* Easier to update tests if needed; otherwise the order will change when adding/removing a single element at the end
* As a user of Dark I would be confused if my output would jump around when adding new elements to a list

It's possible to use a Map and a separate list for storing the order of keys. The issue is that I can't use `Dval` as keys for a Map:
<img width="1130" alt="Screenshot 2023-04-10 at 13 05 52" src="https://user-images.githubusercontent.com/4076804/230881061-367f07cd-95f7-47ee-9e6f-051bc9114886.png">

There is the `equals` function in https://github.com/darklang/dark/blob/main/backend/src/LibExecutionStdLib/LibNoModule.fs#L15
But I would have to create a wrapper type so that it can be used by the map.

I've searched for usages of `Map` in `Lib*.fs` files, but couldn't find a similar usage.

As an alternative, I used a `found` boolean and a `List.foldr` so that there's only a single pass over the list.

Reply:
Cool, good idea.